lib/bundler_api/gem_info.rb

```diff
@@ -65,22 +66,20 @@ def names
     @conn[:rubygems].select(:name).order(:name).all.map {|r| r[:name] }
   end
 
-  # return a list of gem names and their versions
-  def versions
-    specs_hash = Hash.new {|h, k| h[k] = [] }
-    rows = @conn[<<-SQL]
-      SELECT v.full_name, v.number, v.platform
-      FROM versions AS v
-      WHERE v.indexed is true
-SQL
-    rows.each do |row|
-      full_name = row[:full_name]
-      version = row[:number]
-      version += "-" << row[:platform] unless row[:platform] == "ruby"
-      name = full_name.chomp("-" << version)
-      specs_hash[name] << version
+  def versions(date)
+    dataset = @conn[<<-SQL, date]
+          SELECT r.name, v.created_at,
+                 concat_ws('-', v.number, nullif(v.platform,'ruby'))
+          FROM rubygems AS r, versions AS v
+          WHERE v.rubygem_id = r.id AND
+                v.indexed is true AND
+                v.created_at > ?
+    SQL
+    out = {}
+    dataset.each do |entry|
+      out[entry[:name]] ||= []
+      out[entry[:name]] << { number: entry[:concat_ws], created_at: entry[:created_at] }
     end
-
-    specs_hash.each {|k, v| v.sort! }
+    out
```

Comment:
you can use `dataset.each_with_object({})` here to avoid the entire `out` variable